In [1]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [3]:
# Load the dataset (load remotely or locally)
dt = pd.read_csv('./iris.csv')
# Output the first 15 rows of the data
print(dt.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(dt.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

*This is a dataset of 750 samples of flowers, classified into 3 species: setosa, versicolor, and virginica. Each sample has 4 features, its sepal length and width and its petal length and width.*

# Part 2: Split the dataset into train and test

In [7]:
# Take the dataset and split it into our features (X) and label (y)
features = list(dt.keys())
features.remove('species')
X = dt[features]
y = dt['species']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=41)

# Part 3: Logistic Regression

In [8]:
# i. Use sklearn to train a LogisticRegression model on the training set
reg = LogisticRegression()
model = reg.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
probs = model.predict_proba(X_test.iloc[0:1, :])
for p, s in zip(probs[0], model.classes_):
    print(f'Probability for {s}: {p}')
print(f'Correct Label: {y_test.iloc[0]}')
# iii. Report on the score for Logistic regression model, what does the score measure?
print(f'Model score: {model.score(X_test, y_test)}')
# iv. Extract the coefficents and intercepts for the boundary line(s)
print(f'Model Coefficients: {model.coef_}')
print(f'Intercepts: {model.intercept_}')

Probability for setosa: 0.0005466341142886234
Probability for versicolor: 0.5509429753768424
Probability for virginica: 0.448510390508869
Correct Label: virginica
Model score: 0.9333333333333333
Model Coefficients: [[-0.43549552  0.87642112 -2.41999185 -1.0482745 ]
 [ 0.56883363 -0.48546994 -0.2002455  -0.87988876]
 [-0.13333811 -0.39095118  2.62023735  1.92816325]]
Intercepts: [  9.81537952   2.34986154 -12.16524105]


The score measures the prediction accuracy of the model, calculating the percentage of the labels it gets correctly

# Part 4: Support Vector Machine

In [9]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = SVC(probability=True)
model = svc.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
probs = model.predict_proba(X_test.iloc[0:1, :])
for p, s in zip(probs[0], model.classes_):
    print(f'Probability for {s}: {p}')
print(f'Correct Label: {y_test.iloc[0]}')
# iii. Report on the score for the SVM, what does the score measure?
print(f'Model score: {model.score(X_test, y_test)}')

Probability for setosa: 0.008785348537897454
Probability for versicolor: 0.46475948010317336
Probability for virginica: 0.5264551713589293
Correct Label: virginica
Model score: 0.9333333333333333


The score measures the prediction accuracy of the model, calculating the percentage of the labels it gets correctly

# Part 5: Neural Network

In [15]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(max_iter=800, random_state=42) # modified max_iter so that it converges
model = mlp.fit(X_train, y_train)
print(f'Iters: {model.n_iter_}')
# ii. For a sample datapoint, predict the probabilities for each possible class
probs = model.predict_proba(X_test.iloc[6:7, :]) # This one sample datapoint got a wrong prediction
for p, s in zip(probs[0], model.classes_):
    print(f'Probability for {s}: {p}')
print(f'Correct Label: {y_test.iloc[6]}')
# iii. Report on the score for the Neural Network, what does the score measure?
print(f'Model score: {model.score(X_test, y_test)}')

Iters: 543
Probability for setosa: 4.26078738777018e-05
Probability for versicolor: 0.2592604555960311
Probability for virginica: 0.7406969365300913
Correct Label: virginica
Model score: 1.0


When experimenting with `random_state=42` in the train_test_split, the model score was 0.933. After increasing the size of the hidden layer (see below), the model increased to 1.0. Currently, `random_state` is set to `41`.

In [133]:
# iv: Experiment with different options for the neural network, report on your best configuration
mlp = MLPClassifier(hidden_layer_sizes=(3800,), max_iter=1000, random_state=42) # increased the hidden layer size until that datapoint gets predicted correctly
model = mlp.fit(X_train, y_train)
probs = model.predict_proba(X_test.iloc[6:7, :])
for p, s in zip(probs[0], model.classes_):
    print(f'Probability for {s}: {p}')
print(f'Correct Label: {y_test.iloc[6]}')
print(f'Model score: {model.score(X_test, y_test)}')

Probability for setosa: 2.4147394779095824e-06
Probability for versicolor: 0.5409452116764836
Probability for virginica: 0.45905237358403844
Correct Label: versicolor
Model score: 1.0


# Part 6: K-Nearest Neighbors

In [16]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=5)
model = knn.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
probs = model.predict_proba(X_test.iloc[6:7, :]) # This one sample datapoint got a wrong prediction
for p, s in zip(probs[0], model.classes_):
    print(f'Probability for {s}: {p}')
print(f'Correct Label: {y_test.iloc[6]}')
# iii. Report on the score for kNN, what does the score measure?
print(f'Model score: {model.score(X_test, y_test)}')

Probability for setosa: 0.0
Probability for versicolor: 0.4
Probability for virginica: 0.6
Correct Label: virginica
Model score: 0.9333333333333333


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

The MLP classifier performed the best amongst all classifiers, as it gets the highest score with the test/val set. However, for more confidence, cross validation should be used as different splits of the dataset have yielded different results. Nevertheless, with a larger hidden layer size, the MLP classifier was able to achieve a score of 1.0 for every train/test split even if the default setting yielded a lower score. This is because as the size increases, the model increases its complexity, which is different from other classifiers. The performance on actual test set (unseen data) needs to be tested to check if there is any overfit when the hidden layer size increases too much. I don't think anything surprised me.